portico: Add redesigned navbar for landing pages.

amanagr opened the pull request:
This was designed and mostly implemented by @termipost
<img width="1719" alt="image" src="https://user-images.githubusercontent.com/25124304/224909367-7da8cd60-6b20-4c2e-8635-e96bda4bc735.png">


@terpimost it would be great if you can review this.

alya commented:
Excited to see this!

Looks like some links changed from absolute to relative. I don't think that was intended as part of the changes here?

alya commented:
Let's drop the "Demo organization" link for now. We can add a replacement once #24603 is merged.

amanagr commented:
@alya updated based on your feedback. Let me know what you think.

alya commented:
https://github.com/zulip/zulip/pull/24603 just got merged, so we can add a link to it from the top nav now. I'd try putting it last in the left column of the "Product" menu and labeling it "Try Zulip".

alya commented:
> Looks like some links changed from absolute to relative. I don't think that was intended as part of the changes here?

It looks like the links are all relative now, and previously they were all absolute?

terpimost commented:
Looks good. I don't think we need log out icon here 
<img width="143" alt="image" src="https://user-images.githubusercontent.com/1903309/225176052-78d3ebf1-8c3a-4a2c-b374-b256555246d1.png">
but it is not critical.

I've noticed one issue I didn't take care in the past, but I think we should figure it out now or later. There is no way to focus and open submenu with a keyboard. Here is the solution I think would work https://stackoverflow.com/a/63865500 (making label of radio button with tabindex, and producing click). If @amanagr struggle with that, let me know and I try to implement in my example. This could also be a separate pass on the issue

Pricing, Download, Log in and Sign up don't work unless user clicks on the link inside the div, while the whole div should be clickable:
<img width="665" alt="image" src="https://user-images.githubusercontent.com/1903309/225178276-ed6d8555-5eec-48a5-a9a9-def270f0bab0.png">
You can make the whole div as a tag, just be sure styles work the same way 
<img width="624" alt="image" src="https://user-images.githubusercontent.com/1903309/225178359-000d12ba-b531-4524-9819-7da16fbab413.png">

terpimost commented:
submenu close button should also be focusable (requires the same trick as with radio button labels)

amanagr commented:
> It looks like the links are all relative now, and previously they were all absolute?

Opened discussion [here](https://chat.zulip.org/#narrow/stream/107-kandra/topic/relative.20vs.20absolute.20links.20for.20the.20landing.20page)

amanagr commented:
> Looks good. I don't think we need log out icon here

I like the icon there, but I can remove it if others feel like it too.

alya commented:
Removing the icon seems fine to me -- happy to go with Vlad's judgment on this one.

amanagr commented:
Okay

timabbott commented on templates/zerver/landing_nav.html:
```diff
@@ -7,174 +7,252 @@
+                            <rect width="108" height="28" fill="white" />
+                        </clipPath>
+                    </defs>
+                </svg>
```
This seems to change the logo SVG to a different asset; what's the difference? And would it make sense for this to be loaded via an external file or live in the CSS?

amanagr replied:
Yeah, used CSS for it.

timabbott commented on templates/zerver/landing_nav.html:
```diff
@@ -7,174 +7,252 @@
-                </ul>
+                        <a href="/apps">Apps</a>
+                        <a href="https://chat.zulip.org/">Demo organization</a>
+                    </div>
```
I think the same comment I had for the footer applies here -- these are menus that should have a `ul/li` structure, rather than being a bunch of `a` tags inside a `div` with styling to put each `a` tag on its own line.

(And we also want the absolute links, so we're mostly going to revert this to roughly the old HTML structure, albeit with new class names and of course differently organized content)

amanagr replied:
Done.

timabbott commented on templates/zerver/landing_nav.html:
```diff
@@ -7,174 +7,252 @@
+                </label>
+                <div class="top-menu__submenu _right-aligned">
+                    <div class="top-menu__submenu-column">
+                        <a href="/">Got to app</a>
```
"Go to app" looks like it has a typo.

amanagr replied:
Fixed

timabbott commented on templates/zerver/landing_nav.html:
```diff
@@ -7,174 +7,252 @@
+                    <clipPath id="clip0_555_12341234">
+                        <rect width="108" height="28" fill="white" />
+                    </clipPath>
+                </defs>
```
Is this the same logo as above? I guess that would argue for using CSS rather than duplicating its definition.

amanagr replied:
Done

timabbott commented on templates/zerver/landing_nav.html:
```diff
@@ -7,174 +7,252 @@
+        </details>
+        <a href="https://zulip.com/plans/">Pricing</a>
+        <a href="/apps/">Download</a>
+    </div>
```
Is this the exact same content as in the non-mobile view? If so, it'd be really nice if we could avoid duplicating it by just having the difference in visuals for mobile sizes be implemented using CSS media queries rather than having two copies of everything.

(I suspect that's the kind of thing that's annoying to do in the design tool, but ideally we would do in an implementation, because otherwise we have to constantly try to avoid accidental drift between the two copies).

amanagr replied:
Event though the overall content is same, the structure in which it is displayed is very different. So, I don't see a way to  merge them here.

timabbott commented on web/styles/portico/navbar.css:
```diff
@@ -0,0 +1,570 @@
+.top-menu__tab-label-unselect {
+    position: absolute;
+    display: none;
+    right: 15px; /* To account for scrollbar if any present. */
```
Just a nit -- it's almost always nicer to have comments like this on the line before the thing you're describing than inline with it; it usually feels nicer to read the explanation for something before the thing, and you avoid artificial/annoying length constraints for your description.

amanagr replied:
Okay

timabbott commented on web/styles/portico/navbar.css:
```diff
@@ -0,0 +1,570 @@
+    height: 44px;
+    background-size: 32px;
+    background-repeat: no-repeat;
+    background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='rgba(255,255,255,1)' stroke-width='1.5' stroke-linecap='round' stroke-linejoin='round'%3e%3cline x1='18' y1='6' x2='6' y2='18'%3e%3c/line%3e%3cline x1='6' y1='6' x2='18' y2='18'%3e%3c/line%3e%3c/svg%3e");
```
Can you add a comment documenting what this image is?

timabbott commented on web/styles/portico/navbar.css:
```diff
@@ -0,0 +1,570 @@
+
+.top-menu__container {
+    /*   1280px + 32px for paddings on the edges */
+    max-width: 1312px;
```
I think we should try to get into the habit of doing this sort of thing with CSS variables and `calc` calculations, so that it isn't prohibitively difficult to change values like this.

I'm not sure which of the various constants are important to this working, but it'd be great if you can try to extract some of the key constants as variables defined at the top of the file with a comment while this is still fresh. 

(E.g. there's a handful of 3-digit numbers of pixels scattered through this file, and it's a bit hard to guess why any of them have the value they have).

timabbott commented on web/styles/portico/navbar.css:
```diff
@@ -0,0 +1,570 @@
+    font-size: 20px;
+    line-height: 28px;
+    letter-spacing: 0.02em;
+    font-feature-settings: "pnum" on, "lnum" on;
```
@terpimost is it intentional that we're using special font settings for menu items in the navbar?

If so, we should probably add a comment for posterity on why, just because it's just 3 elements in this part of the page that have it. (I think these are settings we used for date/times and might not be intentional here?)

terpimost replied:
I guess it could be a mistake. It should be only for digits. But at this point I'm not sure if we have any numbers in the navbar. It could be for lists?

timabbott replied:
I don't think there are ever any digits here; @amanagr I think you can probably remove the couple lines here that are likely from the timestamp CSS.

amanagr replied:
Done.

timabbott commented on web/styles/portico/navbar.css:
```diff
@@ -0,0 +1,570 @@
+
+.top-menu-mobile[open] {
+    bottom: 0;
+    /* mobile viewport bug fix */
```
Comments like this would be a lot better with more explanation of what the bug is. We almost never want to have "bug fix" be in a comment. Something like "Prevent the mobile viewport from ..." would be a lot more helpful.

amanagr replied:
@terpimost FYI

amanagr replied:
Using `height: 100%` works, well here, I am not sure what @terpimost was trying to fix.

terpimost replied:
I understand what you are trying to tell. if something works, lets use it.

timabbott commented:
Thanks for all the work on this @amanagr and @terpimost! Overall it looks quite nice.

I posted a bunch of comments on the HTML structure; I suspect the `ui/li` part shouldn't be too bad. I don't have a good sense for how much work the duplication of all the menu items for mobile will be to fix; we can certainly consider options like "Having the low-level HTML for the menu items in a separate template that gets included" if it's difficult to do the right thing with CSS. But I can't readily tell whether it is.

I noticed only one visible UI issue playing around with it, which is that the "Pricing/Download/New organization" links don't have the same hover behavior that the ones with a menu do.

(I haven't reviewed the exact arrangement of links and exact labels, since I figure that's easier to fiddle with after the above comments are addressed).

alya commented:
@amanagr Are there any blockers here? If not, it would be great to keep pushing on this.

timabbott commented:
The link added in #24808 should be incorporated here when rebasing, and restored to being https://zulip.com/ rather than a relative link.

alya commented:
> The link added in #24808 should be incorporated here when rebasing, and restored to being [zulip.com](https://zulip.com/) rather than a relative link.

Yep, and we'll need a title for it to match the others. We can try "Managing hundreds of projects at End Point Dev".

amanagr commented:
> @amanagr Are there any blockers here? If not, it would be great to keep pushing on this.

Yeah, I keep thinking today is the day I push an update for this. Maybe today it will be.

amanagr commented on templates/zerver/landing_nav.html:
```diff
@@ -7,174 +7,252 @@
+        </div>
+        <div class='top-menu__item-spacer'></div>
+        <div class='top-menu__items-group-2'>
+            <div class='top-menu__item'><a href="/new">New organization</a></div>
```
Done.

amanagr commented:
Will push an update after reply from @terpimost.

amanagr commented:
I pushed an update but it will be ready for review after Vlad's feedback.

terpimost commented:
I can't test the dev app live, but in chrome dev tools I do see problems. the very bottom part of the footer is cut off and the header gets scrolled:
![CleanShot 2023-03-24 at 00 18 04](https://user-images.githubusercontent.com/1903309/227423332-c50cee32-14ba-4576-85c3-829d823b154b.gif)
In my version footer is cut off too, but header isn't scrollable https://zulip-hello-page.vladkorobov.repl.co/

My original version on a real device (iphone 14) is ok in all browsers (chrome, safari, webview)

terpimost commented:
I also see the same problems with buttons I described earlier 
<img width="281" alt="image" src="https://user-images.githubusercontent.com/1903309/227426201-054aa907-e7d8-4672-8052-d37012672ff0.png">

amanagr commented:
> I can't test the dev app live, but in chrome dev tools I do see problems. the very bottom part of the footer is cut off and the header gets scrolled

I cannot reproduce this in any browser. I not sure what's happening there.

amanagr commented:
I think I addressed all the feedback except the [mobile bug](https://github.com/zulip/zulip/pull/24678#issuecomment-1482239737) reported by Vlad which I cannot reproduce.

amanagr commented:
> Hmm, the nav doesn't seem to be sticky anymore in this verson?

Yeah, that is intentional, since navbar being sticky doesn't look good on all pages. I will try to link you to the discussion after fixing the CI.

amanagr commented:
Oh! The CI is failing due to `/case-studies/end-point/` not being deployed in production yet.

amanagr commented:
Temporarily made /end-point url relative to fix the CI.

alya commented:
> > Hmm, the nav doesn't seem to be sticky anymore in this verson?
> 
> Yeah, that is intentional, since navbar being sticky doesn't look good on all pages. I will try to link you to the discussion after fixing the CI.

Ok, it looked fine to me when I tested, so that would be great.

amanagr commented:
Link to the message concluding the discussion -> https://chat.zulip.org/#narrow/stream/107-kandra/topic/.2Fhello.20.2324301/near/1502315

terpimost commented:
Just noticed some weird border above and below the selected button 
<img width="310" alt="image" src="https://user-images.githubusercontent.com/1903309/227667734-2d829057-ed58-4240-ba95-fca882c149fc.png">

amanagr commented:
@terpimost where are you seeing this?
<img width="854" alt="Screenshot 2023-03-27 at 9 14 06 PM" src="https://user-images.githubusercontent.com/25124304/227992908-a40513d4-c392-4214-af70-e2295c5fde77.png">

timabbott commented:
The main thing I'm worried about here is that the transparent background doesn't look great at a decent portion of scroll positions on some of our pages, so I'm wondering whether we should just figure out a fixed/solid background to use (at least for now) so that we can integrate the rest of the changes.

@alya also noted that "Demo organization" isn't ready yet and should be replaced by a "Try Zulip" link to /try-zulip/

amanagr commented:
> "Demo organization" isn't ready yet and should be replaced by a "Try Zulip" link to /try-zulip/

Done!

amanagr commented:
If we are going to use a solid background, then also we have to worry about the color of it matching the rest of the content in the page.  I am happy to try colors if you want.

Making the navbar non-sticky is another option which would work best right now.

alya commented:
@terpimost proposed [on CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/landing.20page.20nav.20bar.20.2324678/near/1535870):

> yes, lets start from bumping opacity from 0.4 to 0.6

amanagr commented on web/styles/portico/navbar.css:
```diff
@@ -0,0 +1,612 @@
+.top-menu {
+    position: fixed;
+    width: 100%;
+    background: hsl(240deg 100% 10% / 60%);
```
That does look better, you can also look at other opacities by modifying this value.

timabbott replied:
Yeah, this is much improved! I think that unblocks us to merge this as-is and consider revisions if we feel the need for them.

timabbott commented:
This is a great checkpoint, merged, thanks @amanagr!

andersk commented on web/src/portico/header.js:
```diff
@@ -0,0 +1,78 @@
+        on_tab_menu_selection_change();
+    });
+
+    $("body").on("change", "top-menu-tab-input", (e) => {
```
That doesn’t select anything, since it’s (presumably) missing a `.`.
